RoaringBitmap/src/main/java/org/roaringbitmap/art/Node.java

```diff
@@ -0,0 +1,283 @@
+package org.roaringbitmap.art;
+
+import java.io.DataInput;
+import java.io.DataOutput;
+import java.io.IOException;
+
+public abstract class Node {
+
+  //node type
+  protected NodeType nodeType;
+  //length of compressed path(prefix)
+  protected int prefixLength;
+  //the compressed path path (prefix)
+  protected byte[] prefix;
+  //number of non-null children
+  protected int count;
+  public static final int ILLEGAL_IDX = -1;
+
+  public Node(NodeType nodeType, int compressedPrefixSize) {
+    this.nodeType = nodeType;
+    this.prefixLength = compressedPrefixSize;
+    prefix = new byte[prefixLength];
+    count = 0;
+  }
+
+  /**
+   * get the position of a child corresponding to the input key 'k'
+   */
+  public abstract int getChildPos(byte k);
+
+  /**
+   * get the child at the specified position in the node, the 'pos' range from 0 to count
+   */
+  public abstract Node getChild(int pos);
+
+  /**
+   * replace the position child to the fresh one
+   */
+  public abstract void replaceNode(int pos, Node freshOne);
+
+  /**
+   * get the position of the min element in current node.
+   */
+  public abstract int getMinPos();
+
+  /**
+   * get the next position in the node
+   *
+   * @param pos current position
+   */
+  public abstract int getNextLargerPos(int pos);
+
+  /**
+   * get the max child's position
+   */
+  public abstract int getMaxPos();
+
+  /**
+   * get the next smaller element's position
+   */
+  public abstract int getNextSmallerPos(int pos);
+
+  /**
+   * remove the specified position child
+   *
+   * @return an adaptive changed fresh node of the current node
+   */
+  public abstract Node remove(int pos);
+
+  public void serialize(DataOutput dataOutput) throws IOException {
+    serializeHeader(dataOutput);
+    serializeNodeBody(dataOutput);
+  }
+
+  public int serializeSizeInBytes() {
+    int size = 0;
+    size += serializeHeaderSizeInBytes();
+    size += serializeNodeBodySizeInBytes();
+    return size;
+  }
+
+  public static Node deserialize(DataInput dataInput) throws IOException {
+    Node node = deserializeHeader(dataInput);
+    if (node != null) {
+      node.deserializeNodeBody(dataInput);
+      return node;
+    }
+    return null;
+  }
+
+  /**
+   * replace the node's children according to the given children parameter while doing the
+   * deserialization phase.
+   */
+  public abstract void replaceChildren(Node[] children);
+
+  /**
+   * serialize the node's body content
+   */
+  public abstract void serializeNodeBody(DataOutput dataOutput) throws IOException;
+
+  /**
+   * deserialize the node's body content
+   */
+  public abstract void deserializeNodeBody(DataInput dataInput) throws IOException;
+
+  public abstract int serializeNodeBodySizeInBytes();
+
+  /**
+   * find the position that the key and the node's prefix that begins to has a mismatch
+   */
+  public static int prefixMismatch(Node node, byte[] key, int depth) {
```

Comment:
Let's use `Arrays.mismatch` via the Multi-Release shims jar.

Reply:
agree ,will do that